scanner.py

```diff
@@ -1196,22 +1196,25 @@ def get_z_compensation_value(pos):
 
     def _handle_mcu_identify(self):
         try:
-            if self._mcu._mcu_freq < 20000000:
-                self.sensor_freq = self._mcu._mcu_freq
-            elif self._mcu._mcu_freq < 100000000:
-                self.sensor_freq = self._mcu._mcu_freq / 2
-            else:
-                self.sensor_freq = self._mcu._mcu_freq / 6
-            self.inv_adc_max = 1.0 / self._mcu.get_constant_float("ADC_MAX")
-            self.temp_smooth_count = self._mcu.get_constant_float(
-                self.sensor.upper() + "_ADC_SMOOTH_COUNT"
-            )
-            self.thermistor = thermistor.Thermistor(10000.0, 0.0)
-            self.thermistor.setup_coefficients_beta(25.0, 47000.0, 4041.0)
+            if self._mcu.get_constants():
+                if self._mcu._mcu_freq < 20000000:
+                    self.sensor_freq = self._mcu._mcu_freq
+                elif self._mcu._mcu_freq < 100000000:
+                    self.sensor_freq = self._mcu._mcu_freq / 2
+                else:
+                    self.sensor_freq = self._mcu._mcu_freq / 6
+                self.inv_adc_max = 1.0 / self._mcu.get_constant_float("ADC_MAX")
+                self.temp_smooth_count = self._mcu.get_constant_float(
+                    self.sensor.upper() + "_ADC_SMOOTH_COUNT"
+                )
+                self.thermistor = thermistor.Thermistor(10000.0, 0.0)
+                self.thermistor.setup_coefficients_beta(25.0, 47000.0, 4041.0)
 
-            self.toolhead = self.printer.lookup_object("toolhead")
-            self.trapq = self.toolhead.get_trapq()
-            self.fw_version = self._mcu.get_status()["mcu_version"]
+                self.toolhead = self.printer.lookup_object("toolhead")
+                self.trapq = self.toolhead.get_trapq()
+                self.fw_version = self._mcu.get_status()["mcu_version"]
+            else:
+                raise pins.error("Scanner is not connected")
```

Comment:
Invert the condition for an early return - we probably need to also update the type in typings to ensure that it shows that it can return None.
Why is it `pins.error`?


```py
if self._mcu.get_constants() is None:
    raise exception
...
```
Would avoid the indentation from the else block and becomes easier to read.

Reply:
I'm not actually sure why it's pins. That's just how @pellcorp had it setup and working with his beta PR.